### PR TITLE
no-misspelled-operator

### LINKED ISSUE #1: Rule: "=+" should not be used instead of "+="
[RSPEC-2757](https://jira.sonarsource.com/browse/RSPEC-2757)

This rule might intersect with [`whitespace`](https://palantir.github.io/tslint/rules/whitespace/)

### PR DESCRIPTION
**Pull request by inverno**
Fixes #1

### CODE REVIEW

File: tslint.json [hunk 1]
```diff
@@ -15,5 +15,5 @@
         "curly": false,
         "max-line-length": [true, 160]
     },
-    "rulesDirectory": []
+    "rulesDirectory": ["lib/rules/"]
```
Comment (vilchik-elena) [inline tslint.json]: @inverno remove it:)

File: src/rules/noMisspelledOperatorRule.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as tslint from "tslint";
+import * as ts from "typescript";
+import { SonarRuleMetaData } from "../sonarRule";
+import areEquivalent from "../utils/areEquivalent";
+
+export class Rule extends tslint.Rules.AbstractRule {
+  public static metadata: SonarRuleMetaData = {
+    ruleName: "no-misspelt-operator",
```
Comment (vilchik-elena) [inline src/rules/noMisspelledOperatorRule.ts]: @inverno misspelled `misspelt` :-D

File: docs/rules/no-misspelled-operator.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# no-misspelled-operator
+
```
Comment (vilchik-elena) [inline docs/rules/no-misspelled-operator.md]: @inverno we need to update RSPEC accordingly (probably we can do it as JS subtask)

File: src/rules/noMisspelledOperatorRule.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as tslint from "tslint";
+import * as ts from "typescript";
+import { SonarRuleMetaData } from "../sonarRule";
+import areEquivalent from "../utils/areEquivalent";
```
Comment (vilchik-elena) [inline src/rules/noMisspelledOperatorRule.ts]: @inverno unused (ha?! no issue from TSLint?)

File: src/rules/noMisspelledOperatorRule.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    description: "Mispelled operators like '=+', '=-' and '=!' should not be used",
+    optionsDescription: "",
+    options: null,
+    rspecKey: "RSPEC-2757",
+    type: "functionality",
+    typescriptOnly: false,
+  };
+
+  public static formatMessage(identifier: string, suggestedOperator: string) {
+    return `Was "${identifier} ${suggestedOperator}" meant instead?`;
```
Comment (vilchik-elena) [inline src/rules/noMisspelledOperatorRule.ts]: @inverno `x` in message in RSPEC was a placeholder for operator not for variable:)

File: src/rules/noMisspelledOperatorRule.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+ */
+import * as tslint from "tslint";
+import * as ts from "typescript";
+import { SonarRuleMetaData } from "../sonarRule";
+import areEquivalent from "../utils/areEquivalent";
+
+export class Rule extends tslint.Rules.AbstractRule {
+  public static metadata: SonarRuleMetaData = {
+    ruleName: "no-misspelt-operator",
+    description: "Mispelled operators like '=+', '=-' and '=!' should not be used",
```
Comment (vilchik-elena) [inline src/rules/noMisspelledOperatorRule.ts]: @inverno copy to RSPEC

File: src/rules/noMisspelledOperatorRule.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  public apply(sourceFile: ts.SourceFile): tslint.RuleFailure[] {
+    return this.applyWithWalker(new Walker(sourceFile, this.getOptions()));
+  }
+}
+
+class Walker extends tslint.RuleWalker {
+
+  private static COMPOUND_ASSIGNMENT_OPERATORS = new Set([
```
Comment (vilchik-elena) [inline src/rules/noMisspelledOperatorRule.ts]: @inverno `readonly`?

File: src/rules/noMisspelledOperatorRule.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    ts.SyntaxKind.MinusToken,
+    ts.SyntaxKind.ExclamationToken,
+  ]);
+
+  public visitBinaryExpression(node: ts.BinaryExpression) {
+    if (node.operatorToken.kind === ts.SyntaxKind.EqualsToken) {
+      const expression = node.right;
+      if (expression.kind === ts.SyntaxKind.PrefixUnaryExpression) {
+        const unaryExpression = expression as ts.PrefixUnaryExpression;
+        const unaryOperator = unaryExpression.getFirstToken();
```
Comment (vilchik-elena) [inline src/rules/noMisspelledOperatorRule.ts]: @inverno a bit ugly. Did you try to find other ways to retrieve token?

File: src/rules/noMisspelledOperatorRule.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        if (
+          this.isPresentInCompoundAssignments(unaryOperator) &&
+          this.areAdjacent(node.operatorToken, unaryOperator) &&
+          !this.areAdjacent(unaryOperator, unaryExpression.operand)
+        ) {
+          this.addFailure(this.createFailure(
```
Comment (vilchik-elena) [inline src/rules/noMisspelledOperatorRule.ts]: @inverno how about extracting that?

File: tests/rules/noMisspelledOperatorRule/noMisspelledOperatorRule.lint.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+x1 =- 3;
+// ^^  {{Was "x1 =" meant instead?}}
+x1 =+ x1;
+// ^^  {{Was "x1 =" meant instead?}}
+
+x1 =! 2;
+// ^^  {{Was "x1 !=" meant instead?}}
+
+x1 =+4; // OK no whitespace between + and 4
```
Comment (vilchik-elena) [inline tests/rules/noMisspelledOperatorRule/noMisspelledOperatorRule.lint.ts]: @inverno put comment on previous line and generalize it?

File: tests/rules/noMisspelledOperatorRule/noMisspelledOperatorRule.lint.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+x1 =-2; // OK
+x1 =!x1 // OK
+
+let x2 = 0;
```
Comment (vilchik-elena) [inline tests/rules/noMisspelledOperatorRule/noMisspelledOperatorRule.lint.ts]: @inverno new variable?:)

File: src/rules/noMisspelledOperatorRule.ts (hunk 2 above)
Comment (inverno) [inline src/rules/noMisspelledOperatorRule.ts]: It's not misspelt! It's just in a different english :)

File: tslint.json (hunk 1 above)
Comment (inverno) [inline tslint.json]: Done

File: src/rules/noMisspelledOperatorRule.ts (hunk 2 above)
Comment (inverno) [inline src/rules/noMisspelledOperatorRule.ts]: and Done

File: docs/rules/no-misspelled-operator.md (hunk 3 above)
Comment (inverno) [inline docs/rules/no-misspelled-operator.md]: Done

File: src/rules/noMisspelledOperatorRule.ts (hunk 4 above)
Comment (inverno) [inline src/rules/noMisspelledOperatorRule.ts]: @vilchik-elena I like the idea of a type-checker independent version. Create a ticket?

I enabled no-unused-variable and fixed everything it found in the project. So Done

File: src/rules/noMisspelledOperatorRule.ts (hunk 5 above)
Comment (inverno) [inline src/rules/noMisspelledOperatorRule.ts]: Lol, @vilchik-elena thanks a lot! The message indeed didn't look very smart to me and for `=!` it made no sense at all, that's why I put a specific one for that case.

File: src/rules/noMisspelledOperatorRule.ts (hunk 6 above)
Comment (inverno) [inline src/rules/noMisspelledOperatorRule.ts]: Done

File: src/rules/noMisspelledOperatorRule.ts (hunk 7 above)
Comment (inverno) [inline src/rules/noMisspelledOperatorRule.ts]: I didn't even know it existed, thx. Done

File: src/rules/noMisspelledOperatorRule.ts (hunk 5 above)
Comment (inverno) [inline src/rules/noMisspelledOperatorRule.ts]: Done

File: tests/rules/noMisspelledOperatorRule/noMisspelledOperatorRule.lint.ts (hunk 10 above)
Comment (inverno) [inline tests/rules/noMisspelledOperatorRule/noMisspelledOperatorRule.lint.ts]: Done

File: src/rules/noMisspelledOperatorRule.ts (hunk 8 above)
Comment (inverno) [inline src/rules/noMisspelledOperatorRule.ts]: Yeah, nope.

File: tests/rules/noMisspelledOperatorRule/noMisspelledOperatorRule.lint.ts (hunk 11 above)
Comment (inverno) [inline tests/rules/noMisspelledOperatorRule/noMisspelledOperatorRule.lint.ts]: Yeah, I wanted to change

File: src/rules/noMisspelledOperatorRule.ts (hunk 9 above)
Comment (inverno) [inline src/rules/noMisspelledOperatorRule.ts]: Done

**Pull request merged**